src/syscall_sdl.c

```diff
@@ -47,38 +41,75 @@ typedef struct {
 } event_t;
 
 typedef struct {
-    event_t events[EVENT_QUEUE_CAPACITY];
-    size_t start, end;
-    bool full;
+    event_t *base;
+    size_t end;
 } event_queue_t;
 
+enum {
+    RELATIVE_MODE_SUBMISSION = 0,
+};
+
+typedef struct {
+    uint32_t type;
+    union {
+        union {
+            uint8_t enabled;
+        } mouse;
+    };
+} submission_t;
+
+typedef struct {
+    submission_t *base;
+    size_t start;
+} submission_queue_t;
+
 static SDL_Window *window = NULL;
 static SDL_Renderer *renderer;
 static SDL_Texture *texture;
+static uint32_t queues_capacity = 0;
+static uint32_t event_count;
 static event_queue_t event_queue = {
-    .events = {},
-    .start = 0,
+    .base = NULL,
     .end = 0,
-    .full = false,
+};
+static submission_queue_t submission_queue = {
+    .base = NULL,
+    .start = 0,
 };
 
-static bool event_pop(event_t *event)
+static submission_t submission_pop(void)
 {
-    if (event_queue.start == event_queue.end)
-        return false;
-    *event = event_queue.events[event_queue.start++];
-    event_queue.start &= EVENT_QUEUE_CAPACITY - 1;
-    event_queue.full = false;
-    return true;
+    submission_t submission = submission_queue.base[submission_queue.start++];
+    submission_queue.start &= queues_capacity - 1;
+    return submission;
 }
 
-static void event_push(event_t event)
+static void event_push(struct riscv_t *rv, event_t event)
 {
-    if (event_queue.full)
-        return;
-    event_queue.events[event_queue.end++] = event;
-    event_queue.end &= EVENT_QUEUE_CAPACITY - 1;
-    event_queue.full = (event_queue.start == event_queue.end);
+    event_queue.base[event_queue.end++] = event;
+    event_queue.end &= queues_capacity - 1;
+
+    state_t *s = rv_userdata(rv);
+    uint32_t count;
+    memory_read(s->mem, (void *) &count, event_count, sizeof(uint32_t));
+    count += 1;
+    memory_write(s->mem, event_count, (void *) &count, sizeof(uint32_t));
+}
+
+static uint32_t round_pow2(uint32_t arg)
+{
+#ifdef __GNUC__ || __clang__
+    arg = 1 << (32 - __builtin_clz(arg - 1));
+#elif
+    arg--;
```

Comment:
Mention [Bit Twiddling Hacks](https://graphics.stanford.edu/~seander/bithacks.html) for generic implementation.